### PR TITLE
Cache already missed apps in bolt

### PR DESCRIPTION
**Pull request by muymoo**
- Clears missing app bucket every hour

Open questions:
- [x] Should we make user explicitly call `CreateMissingAppsBucket` like we do for the AppBucket?
- [x] Tests? I wasn't sure how to mock cf-client

### CODE REVIEW

**Comment by shinji62** [comment]
I would love to have an opt-in / opt-out option meaning I activate or not this function ... something like that.

**Comment by muymoo** [comment]
@shinji62 When wouldn't you want to ignore missing apps? Would it be enough to configure the ttl of the bucket? Trying to figure out the best way to make it configurable. A flag to enable/disable is kinda messy.

**Comment by muymoo** [comment]
Thoughts on replacing the flag with a TTL option and then only ignoring if TTL is set? I had to change mine locally from 1 hr to 24hrs.

File: caching/caching_boltdb.go [hunk 1]
```diff
@@ -26,22 +32,46 @@ func NewCachingBolt(gcfClientSet *cfClient.Client, boltDatabasePath string) Cach
...
+	c.Appdb.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte(MISSING_APPS_BUCKET))
+		if err != nil {
+			return fmt.Errorf("create bucket: %s", err)
+		}
+		return nil
 	})
 
+	go func() {
+		for range time.Tick(1 * time.Hour) {
```
Comment (shinji62) [inline caching/caching_boltdb.go]: Make this time configurable .

File: caching/caching_boltdb.go [hunk 2]
```diff
@@ -227,15 +231,15 @@ func (c *CachingBolt) GetAppInfoCache(appGuid string) App {
...
 		return App{}
 	}
 
 	// First time seeing app
 	apps := c.GetAppByGuid(appGuid)
-	if apps[0].Name == "" {
+	if c.ignoreMissingApps && apps[0].Name == "" {
 		c.recordMissingApp(appGuid)
-		return App{}
+		return apps[0]
```
Comment (shinji62) [inline caching/caching_boltdb.go]: Not really needed. you return just after.

**Comment by shinji62** [comment]
Agree for the TTL let's use 1h as default.

**Comment by shinji62** [comment]
@muymoo You  are almost done, can you address the last comment ? thanks

**Comment by muymoo** [comment]
Of course! Was out of town this weekend. Should have time this afternoon.

**Comment by muymoo** [comment]
@shinji62 Did you change your mind on the previous comments? I don't see them in this thread anymore.

- Don't use constants
- CreateMissingAppBucket should be used the same way as CreateAppBucket
- Separate function for timer
- Compare to nil instead of .Get()
- Add funcs to interface

**Comment by muymoo** [comment]
@shinji62 Anything else I can do to get merged?

**Comment by shinji62** [comment]
Checking today.

File: main.go [hunk 3]
```diff
@@ -16,26 +16,28 @@ import (
...
-	boltDatabasePath   = kingpin.Flag("boltdb-path", "Bolt Database path ").Default("my.db").OverrideDefaultFromEnvar("BOLTDB_PATH").String()
-	tickerTime         = kingpin.Flag("cc-pull-time", "CloudController Polling time in sec").Default("60s").OverrideDefaultFromEnvar("CF_PULL_TIME").Duration()
-	extraFields        = kingpin.Flag("extra-fields", "Extra fields you want to annotate your events with, example: '--extra-fields=env:dev,something:other ").Default("").OverrideDefaultFromEnvar("EXTRA_FIELDS").String()
-	modeProf           = kingpin.Flag("mode-prof", "Enable profiling mode, one of [cpu, mem, block]").Default("").OverrideDefaultFromEnvar("MODE_PROF").String()
-	pathProf           = kingpin.Flag("path-prof", "Set the Path to write profiling file").Default("").OverrideDefaultFromEnvar("PATH_PROF").String()
+	debug              = kingpin.Flag("debug", "Enable debug mode. This disables forwarding to syslog").Default("false").Envar("DEBUG").Bool()
```
Comment (shinji62) [inline main.go]: What is the issue with OverrideDefaultFromEnvar ?

File: caching/caching_boltdb.go [hunk 4]
```diff
@@ -154,6 +188,33 @@ func (c *CachingBolt) GetAppInfo(appGuid string) App {
...
+func (c *CachingBolt) alreadyMissed(appGuid string) bool {
+	var d []byte
+
+	c.Appdb.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(MISSING_APP_BUCKET))
+		d = b.Get([]byte(appGuid))
+		return nil
+	})
+
+	return bytes.Equal(d, []byte("0"))
```
Comment (shinji62) [inline caching/caching_boltdb.go]: Get  should return nil if not found .. Not sure why you compare like this.

**Comment by shinji62** [comment]
Some small change and should be ok to go.

File: main.go (hunk 3 above)
Comment (muymoo) [inline main.go]: It's deprecated. https://github.com/alecthomas/kingpin/blob/master/flags.go#L255

File: caching/caching_boltdb.go (hunk 4 above)
Comment (muymoo) [inline caching/caching_boltdb.go]: Was trying to be explicit on what we were checking for. Didn't want other values to result in false positives. Changed.

**Pull request merged**